File: VisualIpcvEditor/src/main/java/com/visualipcv/procs/ThresholdProcessor.java
package com.visualipcv.procs;

import com.visualipcv.core.DataBundle;
import com.visualipcv.core.DataType;
import com.visualipcv.core.DataTypes;
import com.visualipcv.core.Processor;
import com.visualipcv.core.ProcessorBuilder;
import com.visualipcv.core.ProcessorProperty;
import org.opencv.core.Mat;
import org.opencv.imgproc.Imgproc;

public class ThresholdProcessor extends Processor {
    public ThresholdProcessor() {
        super(new ProcessorBuilder()
            .setName("Threshold")
            .setModule("Core")
            .setCategory("Image")
            .addInputProperty(new ProcessorProperty("Image", DataTypes.IMAGE))
            .addInputProperty(new ProcessorProperty("Value", DataTypes.DOUBLE))
            .addInputProperty(new ProcessorProperty("Max", DataTypes.DOUBLE))
            .addOutputProperty(new ProcessorProperty("Result", DataTypes.IMAGE)));
    }

    @Override
    public DataBundle execute(DataBundle inputs, DataBundle state) {
        DataBundle result = new DataBundle();
        Mat image = inputs.read("Image");
        Mat out = new Mat();
        Imgproc.threshold(image, out, inputs.read("Value"), inputs.read("Max"), Imgproc.THRESH_BINARY);
        result.write("Result", out);
        return result;
    }
}

File: Native/VisualIPCV/ThresholdProcessor.cpp
//
//  ThresholdProcessor.cpp
//  VisualIPCV
//
//  Created by Артём Новиков on 13.06.2020.
//

#include "ThresholdProcessor.hpp"
#include "ProcessorManager.hpp"

#include <opencv2/core.hpp>
#include <opencv2/imgproc/imgproc.hpp>

using namespace cv;

bool thresholdProcessorProccessorLoadResult = ProcessorManager::registerProcessor(new ThresholdProcessor());

ThresholdProcessor::ThresholdProcessor() : Processor("ThresholdProccessor", "Core", "C++ Image Processors",
{ProcessorProperty("image", BaseDataType(BaseDataTypeClassifier::IMAGE)),
 ProcessorProperty("tresh", BaseDataType(BaseDataTypeClassifier::DOUBLE)),
 ProcessorProperty("max value", BaseDataType(BaseDataTypeClassifier::DOUBLE)),
 ProcessorProperty("treshold type", BaseDataType(BaseDataTypeClassifier::INTEGER))},
{ProcessorProperty("result", BaseDataType(BaseDataTypeClassifier::IMAGE))}) {}

DataBundle ThresholdProcessor::execute(const DataBundle &dataMap, DataBundle &nodeSate) {
    Mat* image;
    try {
        image = dataMap.read<Mat *>(inputProperties[0].name);
    } catch (const std::exception& e) {
        return executionResult(outputProperties[0].name, new Mat());
    }
    double tresh = dataMap.read<double>(inputProperties[1].name);
    double maxValue = dataMap.read<double>(inputProperties[2].name);
    int tresholdType = dataMap.read<int>(inputProperties[3].name);
    
    tresh = tresh < 0 ? 0 : tresh > 255 ? 255 : tresh;
    maxValue = maxValue < 0 ? 0 : maxValue > 255 ? 255 : maxValue;
    tresholdType = tresholdType < 0 ? 0 : tresholdType > 8 ? 16 : tresholdType;
    
    Mat *dstImage = new Mat();
    try {
        threshold(*image, *dstImage, tresh, maxValue, tresholdType);
    } catch (const std::exception& e) {
        return executionResult(outputProperties[0].name, new Mat(*image));
    }
    return executionResult(outputProperties[0].name, dstImage);
}
